2570 Metrics Dashboard and Scan Metrics Widget

Matthew-Grayson opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

- Add Metrics dashboard for admins accessible through Admin Tools.
- Add Scan Metrics widget that displays organization count by http status returned by non-global scans.
- Add scan_result table and models to store http status returned by each organization scanned.
- Add CLI command to populate scan_result table for local development
- Add helper function to write http status to scan_result table and call it from worker.py.
- Add /metrics/scans endpoint to list summarized scan metrics.
- Add /metrics/scans/{scan_id} endpoint to get detailed scan metrics. Add unit tests for /metrics/scans endpoints.
- Add cron job to prune scan_result entries older than 30 days, daily.
- Updated scheduler to upsert total_orgs to the scan table; this is the number of orgs each scan runs against


<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##
See CRASM-2570
When scans fail, developers aren't alerted and we often won't notice failing scans without manual investigation.
Scan metrics allow developers to quickly diagnose failing scans.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##
Added unit tests for each endpoint and ensured existing tests pass.
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->


## 📷 Screenshots (if appropriate) ##
<img width="1792" height="858" alt="Screenshot 2025-08-27 at 10 14 28" src="https://github.com/user-attachments/assets/64bff22e-63e7-4f57-b1fe-ccaf63fa1711" />

<img width="1792" height="858" alt="Screenshot 2025-08-27 at 10 14 44" src="https://github.com/user-attachments/assets/a2dea5cc-b39f-4a87-9abe-04cd5a4b7446" />

<img width="1792" height="849" alt="Screenshot 2025-08-27 at 10 19 08" src="https://github.com/user-attachments/assets/a3acb06c-1563-4d78-8198-4f14d188cea3" />

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Create a release.

Matthew-Grayson committed:
Create ScanResult model for storing timestamp of latest result for each scan/organization pair.

Matthew-Grayson committed:
Merge branch 'develop' of github.com:cisagov/XFD into 2570-scan-metrics

Matthew-Grayson committed:
Merge branch 'develop' of github.com:cisagov/XFD into 2570-scan-metrics

Matthew-Grayson committed:
Update /scans Endpoint to Include Scan Metrics.

Add query to list_scans to count total organizations and organizations that were successfully scanned by each scan.
Add total_orgs and orgs_with_results fields to scan models and interfaces.
Add optional window_days argument to /scans endpoint to specify window for orgs_with_results query.
Update fetchScans call in ScansView.tsx to include window_days.

Matthew-Grayson committed:
Display Scan Metrics in Admin Tools UI.

Add Success Rate column and associated tooltips to Scans page.
Define windowDays state for get /scans endpoint and tooltips.
Increase width of Admin Tools container to better display tables.

Matthew-Grayson committed:
Merge branch 'develop' of github.com:cisagov/XFD into 2570-scan-metrics

Matthew-Grayson committed:
Create helper to track latest_result_at for each scan/org pair.

Matthew-Grayson committed:
Add total_orgs to scan models; modify launch_scan_execution to update total_orgs; improve logging for launch_scan_execution.

Matthew-Grayson committed:
Add Logic to Update scan_result Table When a Scan Saves a Result for an Organization

Rename update_scan_result helper function to upsert_scan_result
Add logic to determine if a scan returns and saves a result for censys, censys_sync, shodan, and shodan_sync
Add logic to upsert timestamp to scan_result table when a scan returns and saves a result for censys, censys_sync, shodan, and shodan_sync

Matthew-Grayson committed:
Add Logic to Assess Success Rate for Non-Global Scans.
 
Add logic to dns_twist.py to update scan_result table if domains are found and saved
Add logic to intel_x_identity.py to update scan_result table if both breaches and credentials are found and saved
Move upsert_scan_result.py from xfd_api.helpers to xfd_api.tasks.helpers

Matthew-Grayson committed:
Add logic to count organizations scanned and track the success of each organization scanned by cybersixgill; add self.scan_id to intel_x __init__.

Matthew-Grayson committed:
Add logic to track success rate for credential_sync scan using scan_result table.

Matthew-Grayson committed:
Modify Metrics Tracking for dns_twist, credential_sync, and sync_asm_sync.

Add logic to credential_sync and sync_asm_sync for recording timestamp of last successful scan per organization.
Add 204 status code to dns_twist for identifying instances where no results are saved.

Matthew-Grayson committed:
Fix Backend Lint Issues.

Refactor Logger.error to use lazy % formatting in censys_sync.py.
Extract queue creation logic from launch_scan_execution to address too_many_statements issue.

Matthew-Grayson committed:
Merge branch 'develop' of github.com:cisagov/XFD into 2570-scan-metrics

Matthew-Grayson committed:
Modify get_scan to Match Updated Scan Model.

Add window_days argument
Add annotation to calculate orgs_with_results
Add error handling to match list_scans method

Matthew-Grayson committed:
Fix Failing Unit Tests.

Add total_orgs and orgs_with_results to scan model in list_scan_tasks.
Extract scan query logic to helper function, query_scans, for use by scan and scan_task api methods.

Matthew-Grayson committed:
Fix Arguments for Scans in views.py.

Fix order for arguments to address lint issues
Consolidate default value for window_days by removing it from ScansView.tsx

Matthew-Grayson committed:
Fix import error related to query_scans.

Matthew-Grayson committed:
Set default value for window_days in view.py.

Matthew-Grayson committed:
Improve ASM Sync, Credential Sync, Shodan, and Scan API Methods.

asm_sync.py:
- Improve error handling and logging
- Add flags to monitor success rate of retrieving both CIDRs and Subdomains
- Track success rate using upsert_scan_result helper function
- Delete commented-out code

credential_sync.py:
- Fix logic for updating data_saved
- Revert code to return status 200 instead of 204 if no data is updated

scan.py
- Modify list_scans and get_scan to return metrics_window_days

shodan.py:
- Improve logging

Matthew-Grayson committed:
Update Logging and Metrics Tracking for Scans.

Update logging for asm_sync
Fix argument order for upsert_scan_results in cybersixgill, dnstwist, and intelx
Update default backend domain for local blocklist queries in dnstwist
Replace f strings in dnstwist
Replace print statements with logging statements in upsert_scan_result helper

Matthew-Grayson committed:
Update Tooltips to Display Metrics Window in Scans View.

Add metrics window to scan response model
Display metrics window in success rate tooltips in scans view
Add tooltip to success rate header in scans view
Removes unused imports and commented out code from scans view

Matthew-Grayson committed:
Add HTTP Status Field to scan_result Table, Model and Helper.
* Add http_status field to ScanResult model.
* Update ScanResult model's unique_together attribute to include scan, organization, and http_status.
* Update scan_result table indexing to improve API speed.
* Update upsert_scan_result helper function to accept and upsert http_status.
* Add http_status to upsert_scan_result log outputs.

Matthew-Grayson committed:
Refactor ScanResult Model and Associated Code to Better Track Scan Failures.

* ScanResult
 - Remove unique_together property on scan, org, http_status.
 - Add message field for error and exception messages.

* upsert_scan_result
 - Remove check for existing scan result.
 - Add optional message argument.
 - Refactor/rename function to log_scan_result.

* asm_sync
 - Add calls to log_scan_result at every exception.

Matthew-Grayson committed:
Update Scan Result Table and Methods.

Rename latest_result_at to scanned_at.
Add ingested_at field and associated index to optimize regular table pruning.
Delete commented out code in models.py.
Add two queries for scan metrics: scan_daily_status_counts and scans_org_count_by_status.
Add api endpoint and response model for scans_org_count_by_status.

Matthew-Grayson committed:
Add scan_result Response Models and Table Pruning Task. 
Write prune_scan_metrics task to delete scan_result entries older than 90 days
Add transformer classes to metrics api methods.
Add GetScanDailyStatusCounts response models.
Remove ingested_at field and index from ScanResult Model.
Prune metrics response models of extra fields.
Write prune_scan_metrics task to delete scan_result entries older than retention period.
Add prune_scan_metrics to SCAN_SCHEMA
Add SCAN_METRICS_RETENTION_PERIOD to example env variables.

Matthew-Grayson committed:
Update list_scans_org_count_by_status to only return metrics for non-global scans.

Matthew-Grayson committed:
Merge branch '2570-scan-metrics' of github.com:cisagov/XFD into 2570-scan-metrics

Matthew-Grayson committed:
Add Metrics Dashboard to Admin Tools.

Add Metrics tab to Admin Tools.
Add @nivo/line library to frontend dependencies.
Add frontend types file for metrics endpoints. Code MetricsDashboard.
Code ScansWidget and tables for scan metrics.

Matthew-Grayson committed:
Add CLI Command to Populate scan_result Table.

Add populate_scan_results function to syncdb_helpers.
Add populate_scan_results CLI command to syncmdl and makefile.

Matthew-Grayson committed:
Refactor prune_scan_results as Cron Job.
 
Remove prune_scan_metrics from SCAN_SCHEMA.
Rename prune_scan_metrics to prune_scan_results.
Schedule pruneScanResults as cron job running every day at midnight.

Matthew-Grayson committed:
Update Metrics API Endpoints. 

Fix cutoff date calculation in metrics apis.
Update get_daily_status_counts to return date string without time.
Fix miscellaneous errors in metrics API methods Remove commented out code from metrics endpoints.
Remove commented out code from metrics endpoints.

Matthew-Grayson committed:
Merge branch 'develop' of github.com:cisagov/XFD into 2570-scan-metrics

Matthew-Grayson committed:
Centralize Logging of Scan Metrics to Worker.

Move call to log_scan_results to worker.py.
Remove calls to log_scan_results from individual scans.
Remove query_scans helper. 
Revert scan endpoints and response_models to no longer accept/return metrics data.
Revert ScansView.tsx to remove metrics window argument.

Matthew-Grayson committed:
Merge branch 'develop' of github.com:cisagov/XFD into 2570-scan-metrics

Matthew-Grayson committed:
Add Tests for Metrics Dashboard Endpoints.

Matthew-Grayson committed:
Merge branch 'develop' of github.com:cisagov/XFD into 2570-scan-metrics

Matthew-Grayson committed:
Fix Lint Issues Flagged by Backend Pipeline / Test.

Matthew-Grayson committed:
Fix Lint Issues Flagged by Backend Pipeline Test.

Matthew-Grayson committed:
Merge branch 'develop' of github.com:cisagov/XFD into 2570-scan-metrics

Matthew-Grayson committed:
Merge branch 'develop' of github.com:cisagov/XFD into 2570-scan-metrics

jayjaybunce commented on backend/src/xfd_django/xfd_api/tasks/helpers/syncdb_helpers/create_sampe_data.py:
Looks like there's a small typo in the file name and across imports from this file: `create_sampe_data` => `create_sample_data`

Matthew-Grayson replied:
I noticed this too. I'm happy to fix this.

Matthew-Grayson replied:
Fixed 83f761043c1ebb8835d7ae1c0065bf9979091371

jayjaybunce reviewed (commented):
LGTM outside typo in file name.

ameliav reviewed (dismissed):
Frontend code LGTM!

cduhn17 commented on backend/src/xfd_django/xfd_api/views.py:
```diff
@@ -495,6 +504,54 @@ async def call_search_logs_filtered(
+    "/metrics/scans",
+    dependencies=[Depends(get_current_active_user)],
+    response_model=ListScansOrgCountByStatusResponse,
+    tags=["metrics"],
```
Is the intended audience of this endpoint to be just for global-admin?

Matthew-Grayson replied:
Yes. Good catch. Working on it now.

Matthew-Grayson replied:
Fixed 83f761043c1ebb8835d7ae1c0065bf9979091371

cduhn17 commented on backend/src/xfd_django/xfd_api/views.py:
```diff
@@ -495,6 +504,54 @@ async def call_search_logs_filtered(
+
+@api_router.get(
+    "/metrics/scans/{scan_id}",
+    dependencies=[Depends(get_current_active_user)],
```
Also is the size/amount of data coming back now or in the future going to be big enough that we may want pagination?

Matthew-Grayson replied:
Good question. We likely won't ever need pagination.
In table one, there is only one row per scheduled scan and I doubt this would ever scale higher than a dozen. In table two there is only one row per unique status code and there are really only a handful of codes that our scans can return.

cduhn17 reviewed (commented):
See comments

Matthew-Grayson committed:
Add globalAdmin Check to Metrics Endpoints; Fix Typos Related to create_sample_data.

Matthew-Grayson commented:
Put in draft mode to optimize designs for 508 compliance.

Matthew-Grayson committed:
Improve Web Accessibility and Design of Metrics Dashboard

Refactor tables to use mui grid to enable keyboard navigation.
Add dynamic aria labels for each column to enable screen reader accessibility.
Update InfoLabel to accept a labelStyle prop to enable using it outside of page headings.

Matthew-Grayson committed:
Merge branch 'develop' of github.com:cisagov/XFD into 2570-scan-metrics

Matthew-Grayson commented:
I refactored the widget to use MUI components and updated styling based on the design team's latest code pushes. I also added dynamic aria labels. These changes make the widget keyboard and screen reader accessible.

cduhn17 reviewed (dismissed):
LGTM

Matthew-Grayson committed:
Merge branch 'develop' of github.com:cisagov/XFD into 2570-scan-metrics

cduhn17 reviewed (approved):
LGTM
